### ISSUE TITLE
Rendering of Nested Serializer fails if no dictionary is parsed

### ISSUE THREAD

**Issue by s4ke** [encode/django-rest-framework#7706]
## Checklist

- [ ] I have verified that that issue exists against the `master` branch of Django REST framework.
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] This is not a usage question. (Those should be directed to the [discussion group](https://groups.google.com/forum/#!forum/django-rest-framework) instead.)
- [x] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](https://www.django-rest-framework.org/community/third-party-packages/#about-third-party-packages) where possible.)
- [x] I have reduced the issue to the simplest possible case.
- [ ] I have included a failing test as a pull request. (If you are unable to do so we can still accept the issue.)

## Steps to reproduce

1. Create a Serializer with a nested Serializer, e.g.:
```
class MySerializer(Serializer):
   payload = PayloadSerializer(default={}, write_only=True)

class PayloadSerializer(Serializer):
  some_field = CharField()
```

2. Post a malformed payload that is not a dictionary:

```
{
   "payload": 1
}
```

## Expected behavior

Nested Serializer should fail and throw a ValidationError, Browsable API should render empty fields.

## Actual behavior

Server crashes in serializer_helpers.py at in as_form_field because integer is not iterable:

```python
    def as_form_field(self):
        values = {}
        for key, value in self.value.items():
            if isinstance(value, (list, dict)):
                values[key] = value
            else:
                values[key] = '' if (value is None or value is False) else force_str(value)
        return self.__class__(self._field, values, self.errors, self._prefix)
```

My workaround, modify Parent serializer:

```python
class PayloadNestedField(NestedBoundField):
    def __init__(self, field, value, errors, prefix=''):
        if value is not None and not isinstance(value, dict):
            value = {}
        super().__init__(field, value, errors, prefix)

class PayloadSerializer(Serializer):
   some_field = CharField()

   def __getitem__(self, key):
            field = self.fields[key]
            value = self.data.get(key)
            error = self.errors.get(key) if hasattr(self, '_errors') else None
            if key == 'payload':
                return PayloadNestedField(field, value, error)
            else:
                return super().__getitem__(key)
```

I think the issue here is that NestedBoundField only defaults the value to {} if it is None or the empty string here:

```python
class NestedBoundField(BoundField):
    """
    This `BoundField` additionally implements __iter__ and __getitem__
    in order to support nested bound fields. This class is the type of
    `BoundField` that is used for serializer fields.
    """

    def __init__(self, field, value, errors, prefix=''):
        if value is None or value == '':
            value = {}
        super().__init__(field, value, errors, prefix)
```

**Comment by tbrknt** [comment]
Hello @s4ke,

I was checking this issue and I tried to reproduce it with the following snippet:
``` python

from rest_framework import serializers 

class PayloadSerializer(serializers.Serializer): 
    some_field = serializers.CharField() 

class MySerializer(serializers.Serializer): 
    payload = PayloadSerializer(default={}, write_only=True) 

data = {'payload': 1}

serializer = MySerializer(data=data)
serializer.is_valid(raise_exception=True)
```
The following Exception is raised:
``` 
ValidationError: {'payload': {'non_field_errors': [ErrorDetail(string='Invalid data. Expected a dictionary, but got int.', code='invalid')]}}
```

I have tested this against the head of [master](https://github.com/encode/django-rest-framework@1ec0f86b585cd87e4b413aeaad1ecc947bacfef2) and I have tested it against the version 3.12.2 that was already installed on my machine.

Do I miss anything in the snippet or in the reproduction steps?

**Comment by s4ke** [comment]
This actually only happens in the browsable UI and is the reason we have not managed to reproduce it in our test suite.

**Comment by tbrknt** [comment]
I understand. Since it's also related with browsable UI, thus the endpoint, would it be possible to give more information regarding the views that can be used to reproduce this? Or does the issue happen with any Viewset extending APIView, GenericView,  ModelViewSet, etc.?

**Comment by s4ke** [comment]
This happens for GenericViewSet with the default html template.

**Comment by tbrknt** [comment]
I have created the following View that uses the serializers from the issue definition. 

``` python
class NestedIssueView(mixins.RetrieveModelMixin, viewsets.GenericViewSet):
    serializer_class = MySerializer
    queryset = Issue.objects.all()

    def testing(self, request, format=None):
        serializer = MySerializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        return Response(serializer.data, status=status.HTTP_201_CREATED)
 
```

I have added it to the `urls.py` as following:
``` python
urlpatterns = [
    path('admin/', admin.site.urls),
    re_path('^issues/', NestedIssueView.as_view({'post': 'testing'}))
]
```

I sent the payload which can be seen in the following screenshot.
![image](https://user-images.githubusercontent.com/6821546/108407681-5aa3b780-7224-11eb-919c-5c12a99b2f38.png)

I am still receiving the validation error.
![image](https://user-images.githubusercontent.com/6821546/108407862-9a6a9f00-7224-11eb-8d12-5ac75e9abe31.png)

Could you maybe give more details to be able to reproduce the issue?

**Comment by s4ke** [comment]
I will try to reproduce the error locally on our codebase and get back to you.

**Comment by s4ke** [comment]
This is the error I get:

![image](https://user-images.githubusercontent.com/719760/108418906-220ada80-7232-11eb-9c50-cbae42ccf8b7.png)

**Comment by tbrknt** [comment]
I was able to reproduce the issue just now. I have tried with `CustomizableBrowsableAPIRenderer`. I was not able to reproduce it. The code was executed in a bit different direction than was in the stacktrace shared above. I was using the following parser class configuration:
```
    'DEFAULT_PARSER_CLASSES': [
        'rest_framework.parsers.JSONParser',
    ],
```

I have changed it to the following:
```
    'DEFAULT_PARSER_CLASSES': [
        'rest_framework.parsers.JSONParser',
        'rest_framework.parsers.FormParser',
        'rest_framework.parsers.MultiPartParser',
    ],
```
Now, I was able to reproduce the issue. Then I have dropped `CustomizableBrowsableAPIRenderer` and used the regular `BrowsableAPIRenderer`. I am still able to reproduce the issue. Thank you for the detailed information. Now I will work on the issue.

**Comment by s4ke** [comment]
Ah, something I also forgot. With this class it works for us:

```
class PayloadNestedField(NestedBoundField):
    def __init__(self, field, value, errors, prefix=''):  # type: ignore
        if value is not None and not isinstance(value, Mapping):
            value = {}
        super().__init__(field, value, errors, prefix)
```

We monkey patch this into the parent serializer class like so:

```
        def __getitem__(self, key):  # type: ignore
            # hack, return a PayloadNestedField to make rendering the browsable API not crash on non dict values
            field = self.fields[key]
            value = self.data.get(key)
            error = self.errors.get(key) if hasattr(self, '_errors') else None
            if key == 'payload':
                return PayloadNestedField(field, value, error)  # type: ignore
            else:
                return super().__getitem__(key)
```

**Comment by tbrknt** [comment]
I have created one PR to reproduce the issue via a test: #7723 

I was checking the options on how to fix it. I believe your solution is the most proper one. Thus, I added it and one test regarding it.
I have created another PR to fix the issue: #7724
